src/js/actions/libraries.js

```diff
@@ -106,13 +111,210 @@ define(function (require, exports) {
+     *
+     * @todo  Eventually, we may need to be better about what to export with the object
+     * and evolve this function as @see models/layer text improves
+     *
+     * @todo Make sure the typeObject is correctly created for everything we're supplying
+     * @todo Update the library itself by either listening to notifications or something
+     *
+     * @return {Promise}
+     */
+    var createCharacterStyleFromSelectedLayer = function () {
+        var appStore = this.flux.store("application"),
+            libStore = this.flux.store("library"),
+            fontStore = this.flux.store("font"),
+            currentDocument = appStore.getCurrentDocument(),
+            currentLibrary = libStore.getCurrentLibrary(),
+            currentLayers = currentDocument.layers.selected,
+            currentLayer = currentLayers.first();
+
+        if (!currentLibrary || currentLayers.count() !== 1 ||
+            !currentLayer || currentLayer.kind !== currentLayer.layerKinds.TEXT) {
+            return Promise.resolve();
+        }
+
+        // FIXME: All constants like this should be described in one location for later
+        var CHARACTERSTYLE_TYPE = "application/vnd.adobe.element.characterstyle+dcx",
+            REPRESENTATION_TYPE = "application/vnd.adobe.characterstyle+json";
+
+        currentLibrary.beginOperation();
+
+        var newElement = currentLibrary.createElement(currentLayer.name, CHARACTERSTYLE_TYPE),
+            representation = newElement.createRepresentation(REPRESENTATION_TYPE, "primary"),
+            imageRepresentation = newElement.createRepresentation("image/png", "rendition"),
+            typeData = fontStore.getTypeObjectFromLayer(currentLayer),
+            // FIXME: Mac/Win temporary locations!
+            filepath = "/tmp/textThumbnailPreview.png";
+
+        // Where magic happens
+        representation.setValue("characterstyle", "data", typeData);
+
+        // FIXME: Make sure this reflects the character style we're recreating
+        // check to see how CEP Panel does it.
+        var exportObj = libraryAdapter.createTextThumbnail(filepath,
+            typeData.adbeFont.postScriptName,
+            "Aa",
+            36,
+            colorAdapter.colorObject([0, 0, 0]));
+
+        return descriptor.playObject(exportObj)
+            .bind(this)
+            .then(function () {
+                return Promise.fromNode(function (cb) {
+                    imageRepresentation.updateContentFromPath(filepath, false, cb);
+                });
+            }).then(function () {
+                // FIXME: Constant here
+                newElement.setRenditionCache(104, filepath, function () {
+                    // FIXME: In CEP Panel, they delete the temporary file afterwards
+                });
+            }).then(function () {
+                currentLibrary.endOperation();
+                // FIXME: Do we need payload?
+                return this.dispatchAsync(events.libraries.ASSET_CREATED, {});
```

Comment:
For future reference, it's fine to just do `this.dispatch` instead of calling the asynchronous version and returning it's promise.